src/meta/src/rpc/server.rs

```diff
@@ -171,308 +158,141 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
-    sub_tasks.push(
-        ClusterManager::start_worker_num_monitor(
-            cluster_manager.clone(),
-            Duration::from_secs(env.opts.node_num_monitor_interval_sec),
-            meta_metrics.clone(),
-        )
-        .await,
-    );
-    sub_tasks.push(HummockManager::start_compaction_heartbeat(hummock_manager).await);
-    sub_tasks.push((election_handle, election_shutdown));
-    if cfg!(not(test)) {
-        sub_tasks.push(
-            ClusterManager::start_heartbeat_checker(cluster_manager, Duration::from_secs(1)).await,
-        );
-        sub_tasks.push(GlobalBarrierManager::start(barrier_manager).await);
-    }
-    let (idle_send, idle_recv) = tokio::sync::oneshot::channel();
-    sub_tasks.push(
-        IdleManager::start_idle_checker(env.idle_manager_ref(), Duration::from_secs(30), idle_send)
-            .await,
-    );
-    let shutdown_all = async move {
-        for (join_handle, shutdown_sender) in sub_tasks {
-            if let Err(_err) = shutdown_sender.send(()) {
-                // Maybe it is already shut down
-                continue;
-            }
-            if let Err(err) = join_handle.await {
-                tracing::warn!("Failed to join shutdown: {:?}", err);
+        let shutdown_all = async move {
+            for (join_handle, shutdown_sender) in svc.sub_tasks {
+                if let Err(_err) = shutdown_sender.send(()) {
+                    // Maybe it is already shut down
+                    continue;
+                }
+                if let Err(err) = join_handle.await {
+                    tracing::warn!("Failed to join shutdown: {:?}", err);
+                }
             }
-        }
-    };
+        };
 
-    let (svc_shutdown_tx, svc_shutdown_rx) = tokio::sync::oneshot::channel::<()>();
+        // FIXME: Add service discovery for leader
+        // https://github.com/risingwavelabs/risingwave/issues/6755
 
-    // Start services.
-    tracing::info!("Starting leader services");
-    let join_handle = tokio::spawn(async move {
         tonic::transport::Server::builder()
-            .layer(MetricsMiddlewareLayer::new(meta_metrics.clone()))
-            .add_service(HeartbeatServiceServer::new(heartbeat_srv))
-            .add_service(ClusterServiceServer::new(cluster_srv))
-            .add_service(StreamManagerServiceServer::new(stream_srv))
-            .add_service(HummockManagerServiceServer::new(hummock_srv))
-            .add_service(NotificationServiceServer::new(notification_srv))
-            .add_service(DdlServiceServer::new(ddl_srv))
-            .add_service(UserServiceServer::new(user_srv))
-            .add_service(ScaleServiceServer::new(scale_srv))
-            .add_service(HealthServer::new(health_srv))
-            .add_service(BackupServiceServer::new(backup_srv))
+            .layer(MetricsMiddlewareLayer::new(svc.meta_metrics))
+            .add_service(HeartbeatServiceServer::new(svc.heartbeat_srv))
```

Comment:
Can we just hide all services in `leader_svs` and return `Router` instead. Then we don't have to store any services in `LeaderServices` struct, which will make the mainline logic more concise.

Reply:
I decided not to return the `Router` but instead start the services  straight away in `leader_svs.rs`. This is done, because I would otherwise need to return `shutdown_all` and other values from the function and returning multiple values is something that we wanted to avoid to simplify the setup. 

The current function definition looks like below. Let me know if you agree with the approach 😄 

```rust 
pub async fn start_leader_srv<S: MetaStore>(
    meta_store: Arc<S>,
    address_info: AddressInfo,
    max_heartbeat_interval: Duration,
    opts: MetaOpts,
    leader_rx: WatchReceiver<(MetaLeaderInfo, bool)>,
    election_handle: JoinHandle<()>,
    election_shutdown: OneSender<()>,
    mut svc_shutdown_rx: WatchReceiver<()>,
) -> MetaResult<()> { ... }
```